marimo/_runtime/reload/autoreload.py

```diff
@@ -261,20 +347,38 @@ def check(
                     continue
                 py_filename, pymtime = module_mtime.name, module_mtime.mtime
 
-                LOGGER.debug(f"Reloading '{modname}'.")
-                try:
-                    superreload(m, self.old_objects)
-                    if py_filename in self.failed:
-                        del self.failed[py_filename]
-                except Exception:
-                    msg = "[autoreload of {} failed: {}]"
-                    LOGGER.debug(
-                        msg.format(modname, traceback.format_exc(10)),
-                    )
-                    self.failed[py_filename] = pymtime
-                else:
-                    # TODO or always evict?
-                    self._module_dependency_finder.evict_from_cache(m)
+                graph = self.get_module_dependency_graph(m)
+                downstream_dependents = get_downstream_dependents(
+                    graph, [get_full_path(m.__file__)]
+                )
+                downstream_dependents_modules = [
+                    modules_paths[dep]
+                    for dep in downstream_dependents
+                    if dep in modules_paths
+                ]
+
+                LOGGER.debug(
+                    f"Reloading '{stale_modname}' and its downstream dependents {downstream_dependents}."
+                )
+                for module in [m] + downstream_dependents_modules:
```

Comment:
Chance of loading the same module multiple times (if `a->b` `c->b` and `a`, `c`, then `b` is called twice

Reply:
I think collecting all outside of the loop, and then just traversing once makes the most sense